plugins/cryptolib/cryptolib.cpp

```diff
@@ -108,11 +108,14 @@ static bool digestsLoaded = false;
 
 static void loadDigests()
 {
-    CriticalBlock block(loadingDigestCS);
     if (!digestsLoaded)
     {
-        OpenSSL_add_all_digests();
-        digestsLoaded = true;
+        CriticalBlock block(loadingDigestCS);
```

Comment:
Unrelated change.
Moving the critical section inside the if test has meant this code is no longer thread safe - to be safe digestsLoaded needs to be an atomic variable.

Reply:
Unfortunately you believe incorrectly...

The problem is that the critical block prevents code from being migrated before the critical block to afterwards, but it *does not* prevent code from being migrated from after the critical block to before it.  Therefore it is perfectly valid for the compiler to generate the following code:

```
    bool temp = !digestsLoaded;
    if (temp)
    {
        CriticalBlock block(loadingDigestCS);
        if (temp)
        {
            OpenSSL_add_all_algorithms();
            digestsLoaded = true;
        }
    }
```

Which still allows the code to be called more than once.  In this situation it is unlikely to cause any problems, but the pattern should be avoided.  Search for "double checked locking unsafe c++" and find more references then you ever want to know about.  The fix in c++11 is to make digestsLoaded an atomic.

One danger of making an unrelated change.